src/authentication/MobileRouter.jsx

```diff
@@ -1,39 +1,119 @@
-import React from 'react'
+import React, { Component } from 'react'
 import { Router, withRouter } from 'react-router'
-
+import Proptypes from 'prop-types'
 import Authentication from './src/Authentication'
 import Revoked from './src/Revoked'
 import { logException } from 'drive/lib/reporter'
+import {
+  readState,
+  secretExchange,
+  getAccessToken,
+  readSecret,
+  clearSecret,
+  clearState,
+  checkExchangedInformations,
+  checkIfOnboardingLogin
+} from './src/utils/onboarding'
+
+class MobileRouter extends Component {
+  async checkState(receivedState, code, cozy_url, history) {
+    const localState = await readState()
+    const localSecret = await readSecret()
+    console.log({ localSecret })
+    console.log({ localState })
+    try {
+      const clientInfo = await secretExchange(
+        localSecret,
+        cozy_url,
+        this.props.client
+      )
+
+      const { onboarding_secret, onboarding_state } = clientInfo
+      console.log({ onboarding_secret })
+      console.log({ onboarding_state })
+      if (
+        !checkExchangedInformations(
+          localSecret,
+          onboarding_secret,
+          localState,
+          onboarding_state
+        )
+      )
+        throw new Error('ERROR', 'exchanged informations are not good')
+
+      const getTokenRequest = await getAccessToken(
+        clientInfo,
+        cozy_url,
+        code,
```

Comment:
i also thought we would get the access token in exchange for an information that only the app and the stack knows, for example the `secret` or the `state`. But here we seem to be using the `code` which can be found in the email? Is this correct? Can we call this route without exchanging secrets first?

Reply:
We are using the `clientInfo` and the clientInfo contains : `onboarding_secret` &  `onboarding_state`

We can call this route without exchanging secret first, but you'll get an error

Reply:
Ok, this is not super obvious from the code but it's the behaviour I expected 👍